crates/light-client-prover/src/circuit.rs

```diff
@@ -61,23 +68,260 @@ pub fn run_circuit<DaV: DaVerifier, G: ZkvmGuest>(
+        last_l2_height = previous_output.last_l2_height;
+        last_state_root = previous_output.state_root;
+        for unchained_info in previous_output.unchained_batch_proofs_info.iter() {
+            // Add them directly as they are the ones that could not be matched
+            initial_to_final.insert(
+                unchained_info.initial_state_root,
+                (
+                    unchained_info.final_state_root,
+                    unchained_info.last_l2_height,
+                ),
+            );
+        }
+
+        for output in deserialized_outputs.iter() {
+            // Do not add if last l2 height is smaller or equal to previous output
+            // This is to defend against replay attacks, for example if somehow there is the script of batch proof 1 we do not need to go through it again
+            if output.last_l2_height <= previous_output.last_l2_height {
+                continue;
+            }
+            recursive_match_state_roots(
+                &mut initial_to_final,
+                &BatchProofInfo::new(
+                    output.initial_state_root,
+                    output.final_state_root,
+                    output.last_l2_height,
+                ),
+            );
+        }
+
+        // Do recursive matching for previous state root
+        recursive_match_state_roots(
+            &mut initial_to_final,
+            &BatchProofInfo::new(
+                previous_output.state_root,
+                previous_output.state_root,
+                previous_output.last_l2_height,
+            ),
+        );
+        // Now only thing left is the state update if exists and others are unchained
+        if let Some((final_root, last_l2)) = initial_to_final.remove(&previous_output.state_root) {
+            last_l2_height = last_l2;
+            last_state_root = final_root;
+        }
+
+        // Collect unchained outputs
+        unchained_outputs = collect_unchained_outputs(&mut initial_to_final);
     }
+    // First light client proof
+    else if let Some(genesis_state_root) = input.l2_genesis_state_root {
+        last_l2_height = 0;
+        last_state_root = genesis_state_root;
 
-    // do what you want with proofs
-    // complete proof has raw bytes inside
-    // to extract *and* verify the proof you need to use the zk guest
-    // can be passed from the guest code to this function
+        for output in deserialized_outputs.iter() {
+            recursive_match_state_roots(
+                &mut initial_to_final,
+                &BatchProofInfo::new(
+                    output.initial_state_root,
+                    output.final_state_root,
+                    output.last_l2_height,
+                ),
+            );
+        }
+
+        // Do recursive matching for previous/genesis state root
+        recursive_match_state_roots(
+            &mut initial_to_final,
+            &BatchProofInfo::new(genesis_state_root, genesis_state_root, 0),
+        );
```

Comment:
what the hell man?  😩
you should insert batch proofs from da to the initial_to_final

and then maybe you can pop recursive